### PR TITLE
feat: [FC-0047] add functionality to send push notifications

### PR DESCRIPTION
**Pull request by NiedielnitsevIvan**
**Description:**
This PR is addressed at adding a new channel that can send push notifications to the OeX mobile app.
A more detailed description can be found in the relevant ADR here - https://github.com/openedx/edx-ace/pull/277

**JIRA:** FC-0047

**Dependencies:** https://github.com/openedx/edx-platform/pull/34971

**Testing instructions:**

1. Install edx-ace version 1.9.0.
2. Checkout the edx-platform from the [implement-push-notifications-chanel](https://github.com/raccoongang/edx-platform/tree/NiedielnitsevIvan/FC-0047/feature/implement-push-notifications-chanel) branch and rebuild.
3. Add FIREBASE_CREDENTIALS or FIREBASE_CREDENTIALS_PATH to the platform settings.
4. Add a new `push_notification` channel to ACE_ENABLED_CHANEL.
5. Log in to OeX.
6. Install the OeX mobile application on your device.
7. Add your device token to FCM Devices with a link to your user from the admin panel.
8. Enroll/unenroll the user(from step 7) in the “Instructor” tab to trigger the push notification event.
9. Make sure you receive push notifications on your device.

**Reviewers:**
- [ ] @e0d 
- [ ] @OmarIthawi 
- [ ] @bmtcril 

**Merge checklist:**
- [ ] All reviewers approved
- [ ] CI build is green
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Commits are squashed

**Post merge:**
- [ ] Create a tag
- [ ] Create a new release (can be done here https://github.com/edx/edx-ace/releases)
- [ ] Check new version is pushed to PyPI after tag-triggered build is 
      finished.
- [ ] Delete working branch (if not needed anymore)

**Author concerns:** 
This PR and the PR that adds the settings for the firebase-admin and django-push-notification package in edx-platform should be merged at the same time to avoid errors.

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @NiedielnitsevIvan!

### What's next?

*Please work through the following steps to get your changes ready for engineering review:*

#### :radio_button: Get product approval

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.

#### :radio_button: Provide context

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads





#### :radio_button: Get a green build

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.



#### :radio_button: Let us know that your PR is ready for review:

### Who will review my changes?


This repository is currently maintained by `@openedx/2u-infinity`.  Tag them in a comment and let them know that your changes are ready for review.


### Where can I find more information?

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)

### When can I expect my changes to be merged?

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*

File: requirements/base.in [hunk 1]
```diff
@@ -7,3 +7,6 @@ attrs>=17.2.0                     # Attributes without boilerplate
 sailthru-client==2.2.3
 six
 stevedore>=1.10.0
+
+firebase-admin==6.5.0
+django-push-notifications==3.1.0
```
Comment (bmtcril) [inline requirements/base.in]: Is there a reason to pin these specific versions? It's different than what's specified in setup.py and we don't usually specify patch versions of things so that we can pick up bug fixes automatically.

**Review by bmtcril — COMMENTED** [review]
Overall the direction of the PR looks good, but please add tests. :) I should be able to run through the testing instructions early next week.

File: edx_ace/ace.py [hunk 2]
```diff
@@ -37,19 +39,36 @@ def send(msg):
...
     msg.report_basics()
 
     channels_for_message = policy.channels_for(msg)
 
     for channel_type in channels_for_message:
+        if limit_to_channels and channel_type not in limit_to_channels:
+            msg.report(
+                'channel_skipped',
+                f'Skipping channel {channel_type}'
+            )
```
Comment (OmarIthawi) [inline edx_ace/ace.py]: If we're using `msg.report` without log_level=INFO or a similar mechanism, I'm concerned that this message would generate log noise given the vast quantity of skipped messages.

I think it's only useful if we use `log_level` or a similar mechanism to differentiates between different message levels for effective log filtering.

File: edx_ace/ace.py [hunk 3]
```diff
@@ -37,19 +39,36 @@ def send(msg):
...
             continue
 
         try:
             rendered_message = presentation.render(channel, msg)
+        except TemplateDoesNotExist as error:
+            msg.report(
+                'template_error',
+                str(error)
+            )
+            continue
```
Comment (OmarIthawi) [inline edx_ace/ace.py]: I see this pattern is used already elsewhere but it's missing logging stacktrace.

File: edx_ace/channel/__init__.py [hunk 4]
```diff
@@ -170,27 +171,29 @@ def get_channel_for_message(channel_type, message):
...
+    except KeyError:
+        return channels_map.get_default_channel(channel_type)
+
+    # First see if any channel specifically demands to deliver this message
+    for channel in possible_channels:
+        if channel.overrides_delivery_for_message(message):
+            return channel
+
+    # Else the normal path: use the preferred channel for this message type
+    return possible_channels[0] if possible_channels else channels_map.get_default_channel(channel_type)
```
Comment (OmarIthawi) [inline edx_ace/channel/__init__.py]: I'm unsure if this is valid anymore to return a single channel for a message.

Would you mind elaborating on why this change was made?

Can we send the same message to both Email and Push?

File: edx_ace/channel/push_notification.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        Transmit a rendered message to a recipient.
+
+        Args:
+            message: The message to transmit.
+            rendered_message: The rendered content of the message that has been personalized
+                for this particular recipient.
+        """
+        device_tokens = self.get_user_device_tokens(message.recipient.lms_user_id)
+        if not device_tokens:
+            LOG.info('Recipient %s has no push token. Skipping push notification.', message.recipient.email_address)
```
Comment (OmarIthawi) [inline edx_ace/channel/push_notification.py]: Is logging `email_address` okay to log?

File: edx_ace/channel/push_notification.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        Get the device tokens for a user.
+        """
+        return list(GCMDevice.objects.filter(
+            user_id=user_id,
+            cloud_message_type='FCM',
+            active=True,
+        ).values_list('registration_id', flat=True))
+
+    @staticmethod
+    def sanitize_html(html_str: str) -> str:
```
Comment (OmarIthawi) [inline edx_ace/channel/push_notification.py]: Why this method is needed?

Also, `sanitize_html` indicates that this method prevents XSS, but its job is mostly to compress spaces. 

```suggestion
    def cleanup_spaces(html_str: str) -> str:
```

or

```suggestion
    def compress_spaces(html_str: str) -> str:
```

File: edx_ace/channel/push_notification.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        }
+        message = dict_to_fcm_message(notification_data)
+        # Note: By default dict_to_fcm_message does not support APNS configuration,
+        #   only Android configuration, so we need to collect and set it manually.
+        apns_config = self.collect_apns_config(notification_data)
+        message.apns = apns_config
+        try:
+            send_message(token, message, settings.FCM_APP_NAME)
+        except Exception as e:
+            LOG.exception('Failed to send push notification to %s', token)
```
Comment (OmarIthawi) [inline edx_ace/channel/push_notification.py:75]: 👍🏼

**Review by OmarIthawi — COMMENTED** [review]
Thanks @NiedielnitsevIvan for this pull request. It's clean and concise.

I have added few questions, please check them and let me know what do you think.

File: edx_ace/renderers.py [hunk 8]
```diff
@@ -81,3 +81,21 @@ class EmailRenderer(AbstractRenderer):
...
+
+
+@attr.s
+class RenderedPushNotification:
+    """
+    Encapsulates all values needed to send a :class:`.Message`
+    over an :attr:`.ChannelType.PUSH`.
+    """
+
+    subject = attr.ib()
```
Comment (OmarIthawi) [inline edx_ace/renderers.py]: I think using `subject` would be confusing since it's an email term.

Using `title` is more appropriate and consistent with the mobile experience.

Sorry it'll require changing many variable and file names, but I hope it's useful.

```suggestion
    title = attr.ib()
```

File: edx_ace/channel/__init__.py (hunk 4 above)
Comment (NiedielnitsevIvan) [reply edx_ace/channel/__init__.py]: This is the `get_channel_for_message` function, which returns only one channel for the received type and message, both in the old implementation and in the new one.

I reworked this to avoid duplicating code when adding a new channel.

This way, messages can be sent via both Email and Push channels at once.

File: edx_ace/renderers.py (hunk 8 above)
Comment (NiedielnitsevIvan) [reply edx_ace/renderers.py]: Changed

File: edx_ace/channel/push_notification.py (hunk 6 above)
Comment (NiedielnitsevIvan) [reply edx_ace/channel/push_notification.py]: Renamed

File: edx_ace/channel/push_notification.py (hunk 5 above)
Comment (NiedielnitsevIvan) [reply edx_ace/channel/push_notification.py]: Changed to user ID in LMS

File: edx_ace/ace.py (hunk 3 above)
Comment (NiedielnitsevIvan) [reply edx_ace/ace.py]: Legging message added.

File: edx_ace/ace.py (hunk 2 above)
Comment (NiedielnitsevIvan) [reply edx_ace/ace.py]: Changed to `log.info`

File: requirements/base.in (hunk 1 above)
Comment (NiedielnitsevIvan) [reply requirements/base.in]: Fixed, and test added.

**Comment by GlugovGrGlib** [comment]
Hi @bmtcril and @OmarIthawi, the questions from the first round of review were fixed and test coverage was increased, PR is ready for the second round of review

**Comment by GlugovGrGlib** [comment]
@bmtcril @OmarIthawi All linters are green except coverage for the patch, I'll try to increase coverage, but I think 95.97% is still good coverage 
Would it be possible to do another round of review this week, so we can unblock the [edx-platform PR](https://github.com/openedx/edx-platform/pull/34971)?

**Comment by bmtcril** [comment]
@GlugovGrGlib I'm working on it now, but it's a non-trivial set of testing. 😄  I'll offer feedback as soon as I can make it work locally, which should hopefully be today.

**Comment by GlugovGrGlib** [comment]
@bmtcril Totally not easy to test it at all. Do you want me to provide you with sandbox environment and testing android or IOS app build, so it's possible to see implemented push notifications in action? This won't provide enough insights in customizability and extendability, but at least will allow to do e2e testing.

**Comment by bmtcril** [comment]
@GlugovGrGlib I've got the app building and a tutor env set up with the right PRs, but are there any docs on the specific oauth settings and firebase side of the setup? I think I can do it from there.

**Comment by GlugovGrGlib** [comment]
@bmtcril Sure, I'll try to collect and provide notes on the configuration!

**Comment by GlugovGrGlib** [comment]
@bmtcril I have created a page with instructions on wiki, latter I'm planning to include it into edx-platform PR, so it will be available on docs.openedx.org
https://openedx.atlassian.net/wiki/spaces/OEPM/pages/4315840517/FC-0047+-+Configure+mobile+push+notifications+in+edx-platform

**Comment by OmarIthawi** [comment]
Thanks for your work. I need more time for review due to travel, and I'll review before the conference.

That being said, please don't block for my review.

**Comment by bmtcril** [comment]
@GlugovGrGlib @NiedielnitsevIvan thanks for the updated instructions, it helped a lot! I think I'm pretty close, but running into a couple of issues:

1- I'm unsure about this step in the testing instructions: "Add your device token to FCM Devices with a link to your user from the admin panel." I don't know what values are necessary for iOS here, it seems like registration ID is a thing we get from Firebase but I haven't been able to find out how.

2- When trying to test sending a message via the push notifications admin panel I'm getting:

`ValueError: The default Firebase app does not exist. Make sure to initialize the SDK by calling initialize_app().` and indeed I don't see any calls to `initialize_app()` in the edx-ace code. Am I missing a setup step?

**Comment by GlugovGrGlib** [comment]
@bmtcril 
1. This device token is generated automatically during registration of the Mobile APP in FCM.
If you're using the latest `develop` branch of the https://github.com/openedx/openedx-app-ios/tree/develop and edx-platform push notifications PR, that include View for device token synchronization - https://github.com/openedx/edx-platform/pull/34971/files#diff-4b3b2718923f3fd9358a6b88e9143a67fd13ca7b74ccdb90f39e3a39c7786757R11, the device token will be pushed to backend [automatically on Login](https://github.com/openedx/openedx-app-ios/pull/461/files#diff-64dbb104fcc6b0b00344cc23aabed82ea80fe678c363d6b53073693f904f1362R32). After successful synchronization you can inspect your device token in admin panel
image
I doubt it's possible to get this device token somehow manually, so you should either enable this view for token synchronization, or you can [log `storage.pushToken`](https://github.com/openedx/openedx-app-ios/pull/461/files#diff-64dbb104fcc6b0b00344cc23aabed82ea80fe678c363d6b53073693f904f1362R31) on your IOS device, and create FCM device in LMS admin panel manually.

2. Regarding FCM initialization please refer to [edx-platform counterpart PR](https://github.com/openedx/edx-platform/pull/34971/files#diff-f694c479e5c9b133241a799e1ddf33d5d5133bfdec91e3f7d371e094c9999e74R36) for firebase app intialization utility. You can either checkout PR branch or just specify these settings and initialize firebase app in private settings.

**Review by bmtcril — APPROVED** [review]
@GlugovGrGlib Got it, I was able to do a successful end-to-end test. Looks good to me, but I think we should wait for another thumb from @OmarIthawi before merging since this isn't an area I'm very familiar with.

File: edx_ace/ace.py [hunk 9]
```diff
@@ -37,19 +43,32 @@ def send(msg):
...
+        limit_to_channels (list of ChannelType, optional): If provided, only send the message over the specified
+            channels. If not provided, the message will be sent over all channels that the policies allow.
     """
     msg.report_basics()
 
     channels_for_message = policy.channels_for(msg)
 
     for channel_type in channels_for_message:
+        if limit_to_channels and channel_type not in limit_to_channels:
+            log.info('Skipping channel %s', channel_type)
```
Comment (OmarIthawi) [inline edx_ace/ace.py]: Perhaps we should lower the log in this case? I'm imagining this will spam logs due to the sheer amount of notifications being sent.

```suggestion
            log.debug('Skipping channel %s', channel_type)
```

File: edx_ace/channel/__init__.py [hunk 10]
```diff
@@ -170,27 +171,29 @@ def get_channel_for_message(channel_type, message):
         Channel: The selected channel object.
     """
     channels_map = channels()
+    channel_names = []
 
     if channel_type == ChannelType.EMAIL:
         if message.options.get('transactional'):
             channel_names = [settings.ACE_CHANNEL_TRANSACTIONAL_EMAIL, settings.ACE_CHANNEL_DEFAULT_EMAIL]
         else:
             channel_names = [settings.ACE_CHANNEL_DEFAULT_EMAIL]
```
Comment (OmarIthawi) [inline edx_ace/channel/__init__.py]:
```suggestion
```

File: edx_ace/channel/__init__.py [hunk 11]
```diff
@@ -170,27 +171,29 @@ def get_channel_for_message(channel_type, message):
...
-        # First see if any channel specifically demands to deliver this message
-        for channel in possible_channels:
-            if channel.overrides_delivery_for_message(message):
-                return channel
-
-        # Else the normal path: use the preferred channel for this message type
-        return possible_channels[0]
-
-    return channels_map.get_default_channel(channel_type)
+    if channel_type == ChannelType.PUSH:
```
Comment (OmarIthawi) [inline edx_ace/channel/__init__.py]: Costmetic changes:

```suggestion
    elif channel_type == ChannelType.PUSH:
```

File: edx_ace/channel/__init__.py (hunk 4 above)
Comment (OmarIthawi) [reply edx_ace/channel/__init__.py]: @NiedielnitsevIvan I'm having a hard time understanding this function and its role after this change. I suppose we may need to have a `get_channel[s]_for_message` function instead.

I understand that it returns a single channel, but I don't think that makes sense anymore given that we send multi-channel messages.

File: edx_ace/tests/channel/test_push_notification.py [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+                lms_user_id=self.lms_user_id,
+            )
+        }
+
+    @override_settings(PUSH_NOTIFICATIONS_SETTINGS={'CONFIG': 'push_notifications.conf.AppConfig'})
+    def test_enabled(self):
+        """
+        Test that the channel is enabled when the settings are configured.
+        """
+        self.assertTrue(PushNotificationChannel.enabled())
```
Comment (OmarIthawi) [inline edx_ace/tests/channel/test_push_notification.py]:
```suggestion
        assert PushNotificationChannel.enabled()
```

File: edx_ace/tests/channel/test_push_notification.py [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+
+    @override_settings(PUSH_NOTIFICATIONS_SETTINGS=None)
+    def test_disabled(self):
+        """
+        Test that the channel is disabled when the settings are not configured.
+        """
+        self.assertFalse(PushNotificationChannel.enabled())
```
Comment (OmarIthawi) [inline edx_ace/tests/channel/test_push_notification.py]:
```suggestion
        assert not PushNotificationChannel.enabled()
```

File: edx_ace/tests/channel/test_push_notification.py [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+        Test that the deliver method sends a push notification for each device token.
+        """
+        with patch.object(PushNotificationChannel, 'get_user_device_tokens', return_value=['token1', 'token2']):
+            message = Message(options={}, **self.msg_kwargs)
+            rendered_message = RenderedPushNotification(title='Test', body='This is a test.')
+
+            channel = PushNotificationChannel()
+            channel.deliver(message, rendered_message)
+
+            self.assertEqual(mock_send_message.call_count, 2)
```
Comment (OmarIthawi) [inline edx_ace/tests/channel/test_push_notification.py]:
```suggestion
            assert mock_send_message.call_count == 2
```

File: edx_ace/tests/channel/test_push_notification.py [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+        """
+        mock_dict_to_fcm_message.return_value = MagicMock()
+        mock_collect_apns_config.return_value = MagicMock()
+
+        message = Message(options={}, **self.msg_kwargs)
+        rendered_message = RenderedPushNotification(title='Test', body='This is a test.')
+
+        channel = PushNotificationChannel()
+
+        with self.assertRaises(FatalChannelDeliveryError):
```
Comment (OmarIthawi) [inline edx_ace/tests/channel/test_push_notification.py]:
```suggestion
        with pytest.raises(FatalChannelDeliveryError):
```

File: edx_ace/tests/channel/test_push_notification.py [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+        """
+        Test that the collect_apns_config method returns an APNSConfig object with the correct headers.
+        """
+        notification_data = {'title': 'Test Title', 'body': 'Test Body'}
+
+        apns_config = PushNotificationChannel.collect_apns_config(notification_data)
+
+        self.assertIsInstance(apns_config, APNSConfig)
+        self.assertEqual(apns_config.headers['apns-priority'], '5')
+        self.assertEqual(apns_config.headers['apns-push-type'], 'alert')
```
Comment (OmarIthawi) [inline edx_ace/tests/channel/test_push_notification.py]:
```suggestion
        assert isinstance(apns_config, APNSConfig)
        assert apns_config.headers['apns-priority'] == '5'
        assert apns_config.headers['apns-push-type'] == 'alert'
```

File: edx_ace/tests/channel/test_push_notification.py [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+
+    def test_get_user_device_tokens(self):
+        """
+        Test that the get_user_device_tokens method returns the device tokens for a user.
+        """
+        gcm_device = GCMDevice.objects.create(user=self.user, registration_id='token1')
+
+        channel = PushNotificationChannel()
+        tokens = channel.get_user_device_tokens(self.lms_user_id)
+        self.assertEqual(tokens, [gcm_device.registration_id])
```
Comment (OmarIthawi) [inline edx_ace/tests/channel/test_push_notification.py]: ditto

File: edx_ace/tests/channel/test_push_notification.py [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+
+    def test_get_user_device_tokens_no_tokens(self):
+        """
+        Test that the get_user_device_tokens method returns an empty list when the user has no device tokens.
+        """
+        channel = PushNotificationChannel()
+        tokens = channel.get_user_device_tokens(self.lms_user_id)
+        self.assertEqual(tokens, [])
```
Comment (OmarIthawi) [inline edx_ace/tests/channel/test_push_notification.py]: same here

File: edx_ace/tests/test_presentation.py [hunk 19]
```diff
@@ -16,7 +15,7 @@ class TestRender(TestCase):
 
     def test_missing_renderer(self):
         channel = Mock(
-            channel_type=ChannelType.PUSH,
+            channel_type='unsupported_channel_type'
```
Comment (OmarIthawi) [inline edx_ace/tests/test_presentation.py:18]: 👍🏼

**Review by OmarIthawi — COMMENTED** [review]
Thanks @NiedielnitsevIvan. I've added many minor changes but my original question is still the same: https://github.com/openedx/edx-ace/pull/282/files#r1650104933

When I saw `possible_channels[0]` it appeared like an anti-pattern as we seamingly pick a random channel and return it.

What would this library look like without the `get_channel_for_message` function?

I think the reason of this confusion is that this function tries to combine a couple concerns into a single place: Return the proper channel for the message given a specific type, but also factor in the `transactional` parameter which complicates the logic a bit.

As a maintainer of this package, I totally support the feature being added and this pull request but I'd like us to take the oppurtunity to identify this bad function design and put a plan to refactor it -- whether now or immediately after the merger of this pull request.

What do you think?

cc: @e0d

File: edx_ace/tests/channel/test_push_notification.py (hunk 12 above)
Comment (OmarIthawi) [reply edx_ace/tests/channel/test_push_notification.py]: `assert` works much better with `pytest` by providing great debugging messages. Please avoid using `self.assert...` unless we have to.

**Comment by NiedielnitsevIvan** [comment]
> @NiedielnitsevIvan I'm having a hard time understanding this function and its role after this change. I suppose we may need to have a get_channel[s]_for_message function instead.
> 
> I understand that it returns a single channel, but I don't think that makes sense anymore given that we send multi-channel messages.

This behavior with `possible_channels[0]` has already been implemented [here](https://github.com/openedx/edx-ace/blob/master/edx_ace/channel/__init__.py#L194C16-L194C36), and I've only added a little refactoring to avoid duplicating code.

File: edx_ace/tests/channel/test_push_notification.py (hunk 18 above)
Comment (NiedielnitsevIvan) [reply edx_ace/tests/channel/test_push_notification.py]: Fixed.

File: edx_ace/tests/channel/test_push_notification.py (hunk 17 above)
Comment (NiedielnitsevIvan) [reply edx_ace/tests/channel/test_push_notification.py]: Fixed.

File: edx_ace/tests/channel/test_push_notification.py (hunk 16 above)
Comment (NiedielnitsevIvan) [reply edx_ace/tests/channel/test_push_notification.py]: Fixed.

File: edx_ace/tests/channel/test_push_notification.py (hunk 15 above)
Comment (NiedielnitsevIvan) [reply edx_ace/tests/channel/test_push_notification.py]: Fixed.

File: edx_ace/tests/channel/test_push_notification.py (hunk 14 above)
Comment (NiedielnitsevIvan) [reply edx_ace/tests/channel/test_push_notification.py]: Fixed.

File: edx_ace/tests/channel/test_push_notification.py (hunk 13 above)
Comment (NiedielnitsevIvan) [reply edx_ace/tests/channel/test_push_notification.py]: Fixed.

File: edx_ace/tests/channel/test_push_notification.py (hunk 12 above)
Comment (NiedielnitsevIvan) [reply edx_ace/tests/channel/test_push_notification.py]: Fixed.

File: edx_ace/channel/__init__.py (hunk 11 above)
Comment (NiedielnitsevIvan) [reply edx_ace/channel/__init__.py]: Fixed.

File: edx_ace/channel/__init__.py (hunk 10 above)
Comment (NiedielnitsevIvan) [reply edx_ace/channel/__init__.py]: Removed.

File: edx_ace/ace.py (hunk 9 above)
Comment (NiedielnitsevIvan) [reply edx_ace/ace.py]: Changed.

File: edx_ace/channel/__init__.py [hunk 20]
```diff
@@ -170,27 +171,28 @@ def get_channel_for_message(channel_type, message):
...
+    except KeyError:
+        return channels_map.get_default_channel(channel_type)
+
+    # First see if any channel specifically demands to deliver this message
+    for channel in possible_channels:
+        if channel.overrides_delivery_for_message(message):
+            return channel
+
+    # Else the normal path: use the preferred channel for this message type
+    return possible_channels[0] if possible_channels else channels_map.get_default_channel(channel_type)
```
Comment (OmarIthawi) [inline edx_ace/channel/__init__.py]: Minor style changes to improve readability.

```suggestion
    if possible_channels:
        return possible_channels[0]
    else: 
        return channels_map.get_default_channel(channel_type)
```

**Review by OmarIthawi — APPROVED** [review]
Thanks @NiedielnitsevIvan for taking the time to answer every review note and your patience for the time it took me to review. 

Admittedly, the original code has confusing logic but it's probably out of the scope of this pull request. This function should have had `if` statements instead of array logic to make it clearer and ease code testing.

I think it makes sense not to block this code due to the logic has been written before. 

Thank you so much, and let me know when you'd like me to merge the pull request.

File: edx_ace/channel/__init__.py (hunk 20 above)
Comment (NiedielnitsevIvan) [reply edx_ace/channel/__init__.py]: Thank you for your review, I have corrected the last comment, so the PR is ready to be merged and released for further use in [this PR](https://github.com/openedx/edx-platform/pull/34971).

**Comment by OmarIthawi** [comment]
Thanks @NiedielnitsevIvan! Please bump up the version to 1.10.0 and I'll squash and merge.

**Comment by NiedielnitsevIvan** [comment]
> Thanks @NiedielnitsevIvan! Please bump up the version to 1.10.0 and I'll squash and merge.

Thanks, ready.

**Comment by GlugovGrGlib** [comment]
@OmarIthawi Done

**Comment by openedx-webhooks** [comment]
@NiedielnitsevIvan 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-ace&entry.867055334=https://github.com/openedx/edx-ace/pull/282&entry.1484655318=https://github.com/NiedielnitsevIvan&entry.752974735=2024-06-11+14:27&entry.1917517419=2024-07-25+16:22&entry.2133058324=Yes) so we can improve your experience in the future.

**Comment by OmarIthawi** [comment]
Thanks @NiedielnitsevIvan @GlugovGrGlib @bmtcril for making this happen.

Code reviews might show the opposite, but I'm really happy to see this getting merged. Only with this feature edX ACE is getting utilized. Otherwise, it was a mostly a useless abstraction 😅
